dm/syncer/safe_mode.go

```diff
@@ -21,11 +21,48 @@ import (
 
 	"github.com/pingcap/tiflow/dm/dm/unit"
 	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/ha"
 )
 
+func (s *Syncer) enableSafeModeByTaskCliArgs(tctx *tcontext.Context) {
+	//nolint:errcheck
+	s.safeMode.Add(tctx, 1)
+	duration, _ := time.ParseDuration(s.cliArgs.SafeModeDuration)
+	duration -= time.Duration(s.tsOffset.Load())
```

Comment:
In my memory we want to calculate safe mode exit time on upstream? So we should always use binlog timestamp to determine exiting safe mode, rather than convert it to  DM's local timeat first. Because there maybe different speed of time passing afterwards.

Reply:
seems this line is not needed
```
	duration -= time.Duration(s.tsOffset.Load())
```

Reply:
i thought this again e.g:

now TS in upsteam : 0 now TS in dm-worker: 1 offset: 1 - 0 = 1

if user input duration: `30s` so they want to exit at TS `0+30` = **30 at upstream**, dm should exit at TS(30 in upsteam 31 at DM )

the wait duration is `now TS in DM  + duration` = `1 + 30` = **31 at DM**

now we can check that if the result are as we expected
binlog with TS(upstream:29,dm:30)  is still in safe mode,but binlog with TS(upstream:30,dm:31) is not in safe mode

---
seems use binlog header TS to check if need to exit safe mode is more direct, i would change the implemention to this way 

Exit safe mode TS = `now TS at DM - offset + duration` = `1 -1 + 30` = `30 at upstream`

Reply:
`SafeModeDuration` is a wall clock duration, why care about time difference between `DM` and `upstream`?

Reply:
It's not the wall clock duration. Because of the import speed keep changing , depend on binlog timestamp is better.
That means:
if the binlog time (DM start parsing the line of binlog) is '00:00:00' and input duration is `30s`. The safe-mode should be exit when parsing the binlog line's timestamp is '00:00:31'.

cc @lance6716 @Ehco1996

Reply:
added in https://github.com/pingcap/tiflow/pull/4601/commits/55cfbd9cec8f129a2d0192345395def0f68a84d6 🧡🧡